cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileStore.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.store.profile;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.Transactionals;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.ConflictDetection;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.api.dataset.table.TableProperties;
+import co.cask.cdap.common.AlreadyExistsException;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.data.dataset.SystemDatasetInstantiator;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.MultiThreadDatasetCache;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.data2.dataset2.lib.table.MetadataStoreDataset;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.profile.Profile;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Store for profile.
+ */
+public class ProfileStore {
+  private static final DatasetId DATASET_ID = NamespaceId.SYSTEM.dataset("profile.meta");
```

Comment:
Is this yet another HBase table? Can we just use the same meta data table as other store?

Reply:
we can use the same app.meta table, though I think it makes sense for this to be a separate class. If we use the same hbase table we should probably move the table name and prefixes to some common class instead of spread throughout several classes.

Reply:
I think there are existing meta store tables using the app.meta name, they just have a unique prefix for them, like https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/MDSStreamMetaStore.java#L59, https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java#L48, and several other places. I agree we should have some classes storing the prefix, it is difficult to know if there is a conflict rowkey if we have many stores using the same name.